### ISSUE TITLE
RFC replace websockets to support http2 and quic

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#2076]
http2 does not provide support to negotiate connection upgrades and therefore does not support websockets [1]. Furthermore reverse proxies suck and are a common cause for false positive "drone is broken" issues due to improper websocket configuration, and in some case (Amazon ELB) don't work at all.

I plan to kill two birds with one stone here.

The recommended solution is to use server sent events (SSE) [2]. Early versions of drone used SSE which I personally preferred to websockets. There were however some issues at the time that made SSE unreliable which is why I replaced the implementation with websockets.

__What has changed with SSE?__

The issue with SSE is that connections frequently timeout after a minute or so due to reverse proxy and load balancer configurations. This required custom configurations (similar to websockets) but in practice had poor support from nginx, caddy and others. The fact that nginx killed a SSE connection normally would not be an issue because SSE will automatically reconnect AND provide the last message ID received so the server can resume.

The problem was, at the time, Drone was unable to seek to a specific location in the stream and resume. This limitation has been removed. This means nginx (et al) can kill connections and the browser will auto-reconnect, and drone can resume with no extra overhead or complexity.

__What about http2 push?__

This would force everyone to use http2 which is not feasible at this time. It would also require generating keys and configuring tls for local development. Bottom line, this would increase the barrier to entry for installing and developing Drone, and would negatively impact my personal development workflow. Server push is also quite new (just landed in Go 1.8). We can therefore re-evaluate http2 server push at a later time.

__What is the end user impact here?__

Absolutely none. This change will take place in the website and server and should not impact usage at all. If you are running Drone on AWS with ELB we probably just made your life way easier. If you are running Drone behind nginx or something else, no changes should be required. Nginx (et al) will certainly timeout connections, but the browser and server will coordinate re-connects and resume streams.

If anything you can expect a performance improvement since we will be able to enable http2 protocol by default if supported by your environment.

[1] https://daniel.haxx.se/blog/2016/06/15/no-websockets-over-http2/
[2] https://www.infoq.com/articles/websocket-and-http2-coexist

**Comment by gtaylor** [comment]
After fighting with SSE on another project last week, the only cautionary tale I have is that you need to be sure to disable any upstream buffering your proxy may be doing. The thing I was seeing was nginx would wait until it had at least partially filled some buffer (or hit a flush interval) before getting back to the end user. The user saw this manifest as a page taking 20+ seconds to "load", whereas nginx was really just not handling the upstream responding with a chunked encoding without a Content-Length.

[This](https://serverfault.com/questions/801628/for-server-sent-events-sse-what-nginx-proxy-configuration-is-appropriate) was super helpful in my diagnosing of the issue.

tldr; You don't need to just worry about timeouts. Buffering is the prime evil.

**Comment by bradrydzewski** [comment]
@gtaylor thanks, that is really good to know. It sounds we can set `X-Accel-Buffering: no`  on the drone server it will prevent nginx from buffering.

**Comment by bradrydzewski** [comment]
The code is now optimized for http2 with the switch from websocket to event source. We also have experimental support for the QUIC protocol. Note that QUIC can only be enabled when using Lets Encrypt.

```
DRONE_LETS_ENCRYPT=true
DRONE_QUIC=true
```

The benefit of QUIC is minimal if your server is located in the same region, but teams who deploy to servers in the US and have team members in EU (for example) should notice a difference.

**Comment by rochdev** [comment]
I know this has unfortunately already been implemented, but I wanted to mention that in my experience websockets with a proper library that includes proper fallbacks is supported pretty much everywhere, and explicitly supported by all reverse proxies, whereas SSEs are supported pretty much nowhere at this point and this in my opinion will cause a lot more problems for drone users than websockets did.

As an example, using the nginx-ingress-controller in Kubernetes (which is very standard) makes it very difficult to support SSE since most of the configuration is managed. This is true especially for routes and IPs which makes https://github.com/drone/docs/issues/230 not viable.

While I agree that in theory SSE are a superior approach to websockets for this use case, in practice websockets have wider support. Since they are already implemented in drone, I would propose maybe supporting both for the moment and letting the users toggle SSE if they prefer.

**Comment by bradrydzewski** [comment]
kubernetes claims to support SSE so you made need to verify you are using the latest version of kubernetes. If you continue to experience issues I recommend reaching out to the kubernetes team
https://github.com/kubernetes/ingress/issues/1091

**Comment by rochdev** [comment]
That PR has been opened 2 days ago and is still not merged though, but I will make sure to follow its progress!

**Comment by bo01ean** [comment]
Do you know if EKS has the version of k8s that has SSE support?

**Comment by gtaylor** [comment]
Wanted to clear up some confusion: This has nothing to do with Kubernetes supporting SSE. The issues linked above were for an nginx-based ingress controller _for_ Kubernetes. Your choice of ingress will dictate support for SSE.
